File: Leetcode 2001-2100/2069/2069.py
# Leetcode 2069: Walking Robot Simulation II
# https://leetcode.com/problems/walking-robot-simulation-ii/
# Solved on 7th of MApril, 2026
class Robot:

    def __init__(self, width: int, height: int):
        self.gridWidth = width
        self.gridHeight = height
        self.perimeter = 2 * (width - 1) + 2 * (height - 1)
        self.currentPos = 0
        self.isStart = True

    def step(self, num: int) -> None:
        self.isStart = False
        self.currentPos = (self.currentPos + num) % self.perimeter

    def getPos(self) -> list[int]:
        if self.currentPos == 0:
            return [0, 0]
        elif self.currentPos < self.gridWidth:
            return [self.currentPos, 0]
        elif self.currentPos < self.gridWidth + self.gridHeight - 1:
            return [self.gridWidth - 1, self.currentPos - self.gridWidth + 1]
        elif self.currentPos < 2 * self.gridWidth + self.gridHeight - 2:
            return [2 * self.gridWidth + self.gridHeight - 3 - self.currentPos, self.gridHeight - 1]
        else:
            return [0, self.perimeter - self.currentPos]

    def getDir(self) -> str:
        if self.currentPos == 0:
            return "East" if self.isStart else "South"
        elif self.currentPos < self.gridWidth:
            return "East"
        elif self.currentPos < self.gridWidth + self.gridHeight - 1:
            return "North"
        elif self.currentPos < 2 * self.gridWidth + self.gridHeight - 2:
            return "West"
        else:
            return "South"

File: Leetcode 2001-2100/2069/2069.java
// Leetcode 2069: Walking Simulation II
// https://leetcode.com/problems/walking-simulation-ii/
// Solved on 7th of April, 2026
class Robot {
    int width;
    int height;
    int perimeter;
    int currentPosition;
    boolean hasMoved;

    public Robot(int width, int height) {
        this.width = width;
        this.height = height;
        this.perimeter = 2 * (width - 1) + 2 * (height - 1);
        this.currentPosition = 0;
        this.hasMoved = false;
    }

    public void step(int num) {
        this.currentPosition = (this.currentPosition + num) % this.perimeter;
        if (num > 0) {
            this.hasMoved = true;
        }
    }

    public int[] getPos() {
        int w1 = this.width - 1;
        int h1 = this.height - 1;

        if (this.currentPosition <= w1) {
            return new int[]{this.currentPosition, 0};
        } else if (this.currentPosition <= w1 + h1) {
            return new int[]{w1, this.currentPosition - w1};
        } else if (this.currentPosition <= 2 * w1 + h1) {
            return new int[]{w1 - (this.currentPosition - (w1 + h1)), h1};
        } else {
            return new int[]{0, h1 - (this.currentPosition - (2 * w1 + h1))};
        }
    }

    public String getDir() {
        if (this.currentPosition == 0) {
            return this.hasMoved ? "South" : "East";
        }
        
        int w1 = this.width - 1;
        int h1 = this.height - 1;

        if (this.currentPosition <= w1) {
            return "East";
        } else if (this.currentPosition <= w1 + h1) {
            return "North";
        } else if (this.currentPosition <= 2 * w1 + h1) {
            return "West";
        } else {
            return "South";
        }
    }
}